pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java

```diff
@@ -0,0 +1,702 @@
+    if (variableSizeDataLength != 0) {
+      _variableSizeDataBytes = new byte[variableSizeDataLength];
+      byteBuffer.position(variableSizeDataStart);
+      byteBuffer.get(_variableSizeDataBytes);
+      _variableSizeData = ByteBuffer.wrap(_variableSizeDataBytes);
+    } else {
+      _variableSizeDataBytes = null;
+      _variableSizeData = null;
+    }
+
+    _trailer = null;
+    /**
+     * V2 stores exceptions as a bunch of KV pairs in metadata, all exceptions has key of "Exception"+errCode.
+     * To interpret V2 bytes as V3 object, extract exceptions from metadata.
+     */
+    _exceptions = extractExceptionsFormV2Metadata();
+  }
+
+  /**
+   * Serialize trailer section to bytes.
+   * Format of the bytes looks:
+   * [numEntries, bytesOfKV2, bytesOfKV2, bytesOfKV3]
+   * For each KV pairs:
+   * - if value is int/long, encode it as: [keyOrdinal, bigEndianRepresentationOfValue]
+   * - if value is string, encode it as: [keyOrdinal, valueLength, Utf8EncodedValue]
+   */
+  private byte[] serializeTrailer()
+      throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
+    int offset = 0;
+    dataOutputStream.writeInt(_trailer.size());
+    offset += Integer.BYTES;
+    for (Map.Entry<TrailerKeys, String> entry : _trailer.entrySet()) {
+      TrailerKeys key = entry.getKey();
+      String value = entry.getValue();
+      dataOutputStream.writeInt(key.ordinal());
+      offset += Integer.BYTES;
+      if (key == TrailerKeys.RESPONSE_SERIALIZATION_CPU_TIME_NS_METADATA_KEY) {
+        _responseSerializationCpuTimeNsValueOffset += offset;
+      }
+      if (IntValueTrailerKeys.contains(key)) {
+        byte[] valueBytes = Ints.toByteArray(Integer.parseInt(value));
+        dataOutputStream.write(valueBytes);
+        offset += valueBytes.length;
+      } else if (LongValueTrailerKeys.contains(key)) {
+        byte[] valueBytes = Longs.toByteArray(Long.parseLong(value));
+        dataOutputStream.write(valueBytes);
+        offset += valueBytes.length;
+      } else {
+        byte[] valueBytes = StringUtil.encodeUtf8(value);
+        dataOutputStream.writeInt(valueBytes.length);
+        dataOutputStream.write(valueBytes);
+        offset += Integer.BYTES + valueBytes.length;
+      }
+    }
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  private Map<TrailerKeys, String> deserializeTrailer(byte[] bytes)
+      throws IOException {
+    try (ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
+        DataInputStream dataInputStream = new DataInputStream(byteArrayInputStream)) {
+      int numEntries = dataInputStream.readInt();
+      Map<TrailerKeys, String> trailer = new TreeMap<>();
+      for (int i = 0; i < numEntries; i++) {
+        int ordinal = dataInputStream.readInt();
+        TrailerKeys key = TrailerKeys.values()[ordinal];
```

Comment:
You should want to check if the oridinal is in range, and skip any unknown ones

Reply:
Now, will throw exceptions for unknown keys. it's ok to throw exception since we don't allow key removal. We can change it as ignore unknown ones if we allow remove keys.

Reply:
We should NEVER throw exception on unknown keys. Please change it so that we always ignore unknown keys. (This has nothing to do with key removal, and let us not bring up key removal again. Key removal is not allowed. Period).